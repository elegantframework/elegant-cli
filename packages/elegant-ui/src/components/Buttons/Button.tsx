import clsx from "clsx";
import React from "react";
import Link from 'next/link';

export interface ButtonProps {
    /**
     * The text to be displayed on the button.
     */
    text: string;
    /**
     * The variant type of button.
     */
    variant?: string;
    /**
     * Custom CSS classes to be applied to the button.
     */
    className?: string;
    /**
     * Provide a url to make this button a Link.
     */
    href?: string;
    /**
     * The primary css color class string.
     */
    primary?: string;
    /**
     * The secondary css color class string.
     */
    secondary?: string;
};

/**
 * A simple button.
 * @returns A pre styled button.
 */
export default function Button({
    text,
    variant="primary",
    className,
    href,
    primary="bg-indigo-800 font-semibold text-indigo-100 hover:bg-indigo-700 active:bg-indigo-800 active:text-indigo-100/70 dark:bg-indigo-700 dark:hover:bg-indigo-600 dark:active:bg-indigo-700 dark:active:text-indigo-100/70",
    secondary="bg-indigo-50 font-medium text-indigo-900 hover:bg-indigo-100 active:bg-indigo-100 active:text-indigo-900/60 dark:bg-indigo-800/50 dark:text-indigo-300 dark:hover:bg-indigo-800 dark:hover:text-indigo-50 dark:active:bg-indigo-800/50 dark:active:text-indigo-50/70",
    ...props
}:ButtonProps) {
    let color = primary;

    if(variant === "secondary") {
        color = secondary;
    }
    
    className = clsx(
        'inline-flex items-center gap-2 justify-center rounded-md py-2 px-3 text-sm outline-offset-2 transition active:transition-none',
        color,
        className
    );

    if(href) {
        return(
            <Link href={href} className={className} {...props}>
                {text}
            </Link>
        );
    }

    return(
        <button className={className} {...props}>
            {text}
        </button>
    );
}