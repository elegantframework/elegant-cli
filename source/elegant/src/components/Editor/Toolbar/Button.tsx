import clsx from 'clsx';
import React from 'react';

export type ButtonVariant = 'primary' | 'secondary' | 'tertiary' | 'quaternary' | 'ghost'
export type ButtonSize = 'medium' | 'small' | 'icon' | 'iconSmall'

export type ButtonProps = {
  variant?: ButtonVariant
  active?: boolean
  activeClassname?: string
  buttonSize?: ButtonSize
} & React.ButtonHTMLAttributes<HTMLButtonElement>

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    { active, buttonSize = 'medium', children, disabled, variant = 'primary', className, activeClassname, ...rest },
    ref,
  ) => {
    const buttonClassName = clsx(
      'flex group items-center justify-center border border-transparent gap-2 text-sm font-semibold rounded-md disabled:opacity-50 whitespace-nowrap',

      variant === 'primary' &&
        clsx(
          'text-white bg-black border-black dark:text-black dark:bg-white dark:border-white',
          !disabled &&
            !active &&
            'hover:bg-neutral-800 active:bg-neutral-900 dark:hover:bg-neutral-200 dark:active:bg-neutral-300',
          active && clsx('bg-neutral-900 dark:bg-neutral-300', activeClassname),
        ),

      variant === 'secondary' &&
        clsx(
          'text-neutral-900 dark:text-white',
          !disabled &&
            !active &&
            'hover:bg-neutral-100 active:bg-neutral-200 dark:hover:bg-neutral-900 dark:active:bg-neutral-800',
          active && 'bg-neutral-200 dark:bg-neutral-800',
        ),

      variant === 'tertiary' &&
        clsx(
          'bg-neutral-50 text-neutral-900 dark:bg-neutral-900 dark:text-white dark:border-neutral-900',
          !disabled &&
            !active &&
            'hover:bg-neutral-100 active:bg-neutral-200 dark:hover:bg-neutral-800 dark:active:bg-neutral-700',
          active && clsx('bg-neutral-200 dark:bg-neutral-800', activeClassname),
        ),

      variant === 'ghost' &&
        clsx(
          'bg-transparent border-transparent text-neutral-500 dark:text-neutral-400',
          !disabled &&
            !active &&
            'hover:bg-black/5 hover:text-neutral-700 active:bg-black/10 active:text-neutral-800 dark:hover:bg-white/10 dark:hover:text-neutral-300 dark:active:text-neutral-200',
          active && clsx('bg-black/10 text-neutral-800 dark:bg-white/20 dark:text-neutral-200', activeClassname),
        ),

      buttonSize === 'medium' && 'py-2 px-3',
      buttonSize === 'small' && 'py-1 px-2',
      buttonSize === 'icon' && 'w-8 h-8',
      buttonSize === 'iconSmall' && 'w-6 h-6',

      className,
    )

    return (
      <button ref={ref} disabled={disabled} className={buttonClassName} {...rest}>
        {children}
      </button>
    )
  },
)

Button.displayName = 'Button'
