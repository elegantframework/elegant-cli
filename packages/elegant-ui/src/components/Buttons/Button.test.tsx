import React from "react";
import renderer from 'react-test-renderer';
import Button from './Button';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

describe("Button component", () => {
    it('renders the button html as expected', () => {
        const element = renderer
        .create(
            <Button text="Hello World"/>
        )
        .toJSON();
        expect(element).toMatchSnapshot();
    });

    it('displays the button text as expected', () => {
        render(
            <Button text="Hello World"/>
        );

        expect(
            screen.getByText('Hello World')
        ).toBeInTheDocument();
    });

    it('renders a link when passed an href', () => {
        render(
            <Button text="Hello World" href="example.com"/>
        );

        expect(
            screen.getByText('Hello World')
        ).toBeInTheDocument();
    });

    it('renders the secondary variant as expected', () => {
        const element = renderer
        .create(
            <Button text="Hello World" variant="secondary"/>
        )
        .toJSON();
        expect(element).toMatchSnapshot();
    }) 
});