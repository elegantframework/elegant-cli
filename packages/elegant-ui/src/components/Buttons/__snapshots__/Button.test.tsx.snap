// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders the button html as expected 1`] = `
<button
  className="inline-flex items-center gap-2 justify-center rounded-md py-2 px-3 text-sm outline-offset-2 transition active:transition-none bg-indigo-800 font-semibold text-indigo-100 hover:bg-indigo-700 active:bg-indigo-800 active:text-indigo-100/70 dark:bg-indigo-700 dark:hover:bg-indigo-600 dark:active:bg-indigo-700 dark:active:text-indigo-100/70"
>
  Hello World
</button>
`;

exports[`Button component renders the secondary variant as expected 1`] = `
<button
  className="inline-flex items-center gap-2 justify-center rounded-md py-2 px-3 text-sm outline-offset-2 transition active:transition-none bg-indigo-50 font-medium text-indigo-900 hover:bg-indigo-100 active:bg-indigo-100 active:text-indigo-900/60 dark:bg-indigo-800/50 dark:text-indigo-300 dark:hover:bg-indigo-800 dark:hover:text-indigo-50 dark:active:bg-indigo-800/50 dark:active:text-indigo-50/70"
>
  Hello World
</button>
`;
